Updated Handling of QM Atoms

ape33 opened the pull request:
This PR improves the internal handling of QM atoms and prepares for QM/MM approaches

Commit:
internal: add getQMAtomsNew() and qmmm attributes to atoms class

Commit:
bugfix: moved getQMAtomsNew() to .hpp file
this has to be done since:
1. auto keyword has be the return type bc of lambda
2. full function def needs to be seen when auto is used
3. inline added to comply with ODR

ape33 committed:
internal: replaced getQMAtoms in dftbplusRunner

ape33 committed:
internal: lower number of mem allocations in dftbplusRunner

ape33 committed:
internal: dftbplusRunner improved uniqueatomnames

ape33 committed:
internal: updated getQMAtoms in pyscf runner

ape33 committed:
internal: updated getQMAtoms in turbomoleRunner runner

ape33 committed:
internal: getQMAtomsNew for all qm-type jobs

ape33 committed:
internal: removed getuniqueqmatomnames and qm_atoms in copy of sim box

ape33 committed:
internal: removed _qmAtoms from simBox

ape33 committed:
internal: renaming of isQMOnlyJobtype()

ape33 committed:
internal: _isQMOnly renamed to _isForcedQM

ape33 committed:
internal: renamed getQMAtomsNew() to getQMAtoms()

ape33 committed:
internal: removed _qmCenterAtoms

ape33 committed:
internal: renamed flattenPositions()

ape33 committed:
internal: aseQMRunner now acts on qmatoms only

97gamjak commented on src/QM/external/dftbplusRunner.cpp:
```diff
@@ -64,66 +63,72 @@ void DFTBPlusRunner::writeCoordsFile(SimulationBox &box)
+    coordsFile << "  " << (Settings::getJobtype() == JobType::QM_MD ? 'S' : 'C')
+               << '\n';
 
-    const auto uniqueAtomNames = box.getUniqueQMAtomNames();
+    std::set<std::string> uniqueAtomNames;
+    for (const auto &atom : box.getQMAtoms())
+        uniqueAtomNames.insert(atom->getName());
```
why did you remove this member function from the simulationbox class? IMO it makes more sense to have it there!

97gamjak replied:
the set appraoch is a really nice idea, however i would really put this also in the simbox member funciton

Lucky-Armin replied:
I removed the `getUniqueQMAtomNames()` member function, because the original implementation had more lines of code and was only used once (here in the `dftbplusRunner.cpp`). But yes, I can put this functionality inside a member function and call it here.

97gamjak commented on src/QM/external/pyscfRunner.cpp:
```diff
@@ -24,9 +24,9 @@
-#include <format>    // for format
-#include <fstream>   // for ofstream, operator<<, basic_ostream
-#include <string>    // for allocator, string, operator+, operator<<
+#include <format>     // for format
+#include <fstream>    // for ofstream, operator<<, basic_ostream
+#include <string>     // for allocator, string, operator+, operator<<
```
why? formatting issue again?

97gamjak commented on src/simulationBox/simulationBox.cpp:
```diff
@@ -153,12 +147,14 @@ void SimulationBox::setupQMOnlyAtoms(const std::vector<int>& atomIndices)
-
-        if (it == _qmAtoms.end())
-            _qmAtoms.push_back(_atoms[(size_t) index]);
+        if (_atoms[(size_t) index]->isForcedMM())
```
don't use c-style casts! i know it comes from the previous implementation but at this point we can correct this!

Lucky-Armin replied:
I totally agree! I only used it, because it was like this before...

97gamjak reviewed (changes requested):
Okay so i have to global comments about the current PR:

1) i would strongly suggest to directly go to a logic in atoms.hpp where we define everything with the hybrid approach so core instead of qm core and so on.
You can leave all implementation details in the cpp files as they are right now. This means you don't have to take into account errors if the hybrid approach isn't qmmm and so on. To change this you would of course have to rewrite the getqmatoms call to check if it is either a qmonly approach or a qmmm approach. Let me know if anything is unclear with this.

2) We need integration tests!!! can you please try to setup some integration tests for all qm methods you changed in the qmmm-dev branch? so ase pyscf dftb... because the things we introduce here are really breaking changes so we have to be 100% sure that everything works as expected. For this you can check the deleted integration tests of dftbplus some time ago from dev and master. Plus for the moment you can only try to make some integration tests that work locally for dev and feature branch and I help you with the CI integration.

Lucky-Armin commented on include/simulationBox/simulationBox.hpp:
```diff
@@ -186,15 +185,14 @@ namespace simulationBox
 
         [[nodiscard]] pq::SharedAtomVec&         getAtoms();
-        [[nodiscard]] pq::SharedAtomVec&         getQMAtoms();
+        [[nodiscard]] auto                       getQMAtoms() const;
```
The main reason for the `auto` keyword here was that there is no explicit type for a lambda function (which is part of the return type).
I'll take a look how to handle it in another way.

ape33 committed:
internal: added getUniqueQMAtomNames()

ape33 committed:
internal: using declarations in dftbplusRunner

Lucky-Armin commented on src/QM/external/dftbplusRunner.cpp:
```diff
@@ -64,66 +63,72 @@ void DFTBPlusRunner::writeCoordsFile(SimulationBox &box)
-
-        const auto iter   = find(uniqueAtomNames, atom.getName());
-        const auto atomId = distance(uniqueAtomNames.begin(), iter) + 1;
+        const auto iter = std::ranges::find(uniqueAtomNames, atom->getName());
+        const auto atomId =
+            std::ranges::distance(uniqueAtomNames.begin(), iter) + 1;
```
The reason I removed it was that it was only used once. However, I didn't realize that the line using `std::ranges::distance` now spans two lines. I put the using declarations back in. However, not at the top, but at the beginning of the function. I hope this is ok.

ape33 committed:
internal: fixed formatting in pyscfRunner

ape33 committed:
internal: removed C-style casts in simulationBox.cpp

ape33 committed:
test: added integration tests for ase_dftb and xtb

ape33 committed:
test: updated ref_data for ase runs

ape33 committed:
test: added integration tests for dftb+ and mace

ape33 committed:
test: added integration tests for pyscf

ape33 committed:
test: added dependecies for pyscf integration test

Commit:
test: added dependecies for dftb+ integration test

Commit:
test: add conda to system path for ci

Commit:
test: updated dependencies for ci

Commit:
test: updated dependencies for ci

ape33 committed:
test: setup conda env in ci workflow

ape33 committed:
test: added dependencies for mace integration test

ape33 committed:
test: reorder ci actions

ape33 committed:
test: removed integration test for ubuntu arm

ape33 committed:
test: removed integration_test for mace

ape33 committed:
test: updated ci workflow order

ape33 committed:
test: re-added arm integration tests

ape33 committed:
test: ci integration test dependencies fixed

ape33 committed:
internal: created helper function isQMAtom()

Lucky-Armin commented on include/simulationBox/simulationBox.hpp:
```diff
@@ -248,6 +246,35 @@ namespace simulationBox
+                                else if (atom->isForcedQM())
+                                    return true;
+
+                                return false;
```
I created a helper function `Atom::isQMAtom()`

ape33 committed:
internal: renamed qmmm atom properties to hybrid

ape33 committed:
internal: consider all qm only jobtypes in dftbplusRunner

ape33 committed:
internal: rephrasing and reformatting

Commit:
ci: comment for arm dftbplus build

Commit:
Merge branch 'qmmm-dev' into getQMAtoms

ape33 committed:
workflow: specify pyscf version to get correct F

ape33 committed:
feature: added qm script for pyscf hf calc

ape33 committed:
test: remove force chk for pyscf integration test

97gamjak committed:
feat: started implementing simulationbox class view

Commit:
internal: removed HybridType for atoms class

Commit:
internal: added _isActive to Atom class

Commit:
doc: updated doxygen comment

ape33 committed:
internal: add getMMAtoms() and isMMAtom()

ape33 committed:
internal: renamed QMonly, MMonly to forcedInner, forcedOuter

ape33 committed:
internal: renamed qmCenter to innerRegionCenter

ape33 committed:
internal: renamed qmmmSetup to hybridSetup

ape33 committed:
internal: add isHybridJobtype()

ape33 committed:
internal: add jobtype filter for isQMAtom() & isMMAtom()

ape33 committed:
internal: removed C-style cast

ape33 committed:
internal: qmRunner - added dimensionality

ape33 committed:
internal: added _periodicity to QM Runners

Commit:
merge: remove integration_tests to resolve conflict

Commit:
merge: remove ci_build to resolve conflict

Commit:
workflow: readded ci_build

Commit:
Merge branch 'qmmm-dev' into getQMAtoms

Commit:
tests: re-added integration_tests

97gamjak committed:
Merge remote-tracking branch 'origin/getQMAtoms' into crtp-simbox-view

97gamjak committed:
Add filter adaptor and view implementation

- Introduced FilterAdaptor as a functor for creating FilterView instances based on a predicate.
- Implemented FilterView to provide a range-based view that filters elements according to a given predicate.
- Developed FilterIterator and FilterSentinel to facilitate iteration over filtered ranges.
- Added operator| to enable seamless application of adaptors to ranges.
- Included necessary headers and ensured compliance with the GNU General Public License.

97gamjak committed:
cleanup: for PR

Lucky-Armin commented on include/settings/hybridSettings.hpp:
```diff
@@ -54,9 +54,15 @@ namespace settings
-        static void setCoreOnlyListString(const std::string_view qmOnlyList);
-        static void setNonCoreOnlyListString(const std::string_view mmOnlyList);
+        static void setInnerRegionCenterString(
+            const std::string_view innerRegionCenter
```
&

Lucky-Armin commented on src/settings/settings.cpp:
```diff
@@ -257,7 +257,41 @@ bool Settings::isQMOnly()
+{
+    using enum JobType;
+
+    if (_jobtype == MM_MD)
+        return true;
+
+    else
+        return false;
```
return ... == ...

97gamjak committed:
refactor: cleaning up PR

ape33 committed:
refactor: add getBoxMatrix with periodicity

97gamjak committed:
Merge pull request #178 from MolarVerse/crtp-simbox-view

Crtp simbox view

97gamjak committed:
refactor: remove unused readForces function in qmrunner

97gamjak committed:
Update include/settings/hybridSettings.hpp

Co-authored-by: Armin Penz <129873177+Lucky-Armin@users.noreply.github.com>

ape33 committed:
internal: added missing &

ape33 committed:
test: add settings test

ape33 committed:
internal: added transform view for getQMAtomicNumbers

97gamjak commented on include/views/transformSentinel.hpp:
Please use a common sentinel for both iterators as they are exactly the same

97gamjak commented on include/views/transformIterator.hpp:
```diff
@@ -0,0 +1,166 @@
+
+        reference          operator*() const;
+        TransformIterator& operator++();
+        TransformIterator  operator++(int);
```
Why did you decide to not implement the address of operator? It is kind of not that straight forward here, so we can leave it for later but we should keep it in mind.

97gamjak reviewed (changes requested):
FYI: I won't ask for a lot of simplifications or generalization in the views for now becuase I plan there something else for the future. But the Sentinel shold be really done.

ape33 committed:
refactor: create generic sentinel for custom iter
